roachtest: sqlsmith nil panic due to incorrect sequence desc resolution

[(roachtest).sqlsmith/setup=tpcc/setting=no-mutations failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2653401&tab=buildLog) on [master@5b33e6dfc47000de831745a851e3bf9e2cf7fd95](https://github.com/cockroachdb/cockroach/commits/5b33e6dfc47000de831745a851e3bf9e2cf7fd95):

```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/sqlsmith/setup=tpcc/setting=no-mutations/run_1
	sqlsmith.go:198,sqlsmith.go:228,test_runner.go:767: error: pq: internal error: unexpected error from the vectorized engine: runtime error: invalid memory address or nil pointer dereference
		stmt:
		SELECT
			tab_872.tableoid AS col_1949,
			tab_873.o_all_local AS col_1950,
			tab_872.crdb_internal_mvcc_timestamp AS col_1951,
			4912713965629465434:::INT8 AS col_1952,
			tab_873.o_carrier_id AS col_1953,
			nextval(tab_872.tableoid::REGCLASS)::INT8 AS col_1954,
			tab_872.crdb_internal_mvcc_timestamp AS col_1955,
			NULL AS col_1956,
			8464184514683886707:::INT8 AS col_1957,
			NULL AS col_1958,
			'-4713-11-24 00:00:00':::TIMESTAMP AS col_1959,
			tab_872.tableoid AS col_1960,
			tab_872.no_o_id AS col_1961
		FROM
			defaultdb.public.new_order@[0] AS tab_872, defaultdb.public."order"@[0] AS tab_873
		ORDER BY
			tab_872.tableoid ASC;

```

<details><summary>More</summary><p>


Artifacts: [/sqlsmith/setup=tpcc/setting=no-mutations](https://teamcity.cockroachdb.com/viewLog.html?buildId=2653401&tab=artifacts#/sqlsmith/setup=tpcc/setting=no-mutations)

[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2Asqlsmith%2Fsetup%3Dtpcc%2Fsetting%3Dno-mutations.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

--- a/pkg/sql/logictest/testdata/logic_test/sequences
+++ b/pkg/sql/logictest/testdata/logic_test/sequences
@@ -1394,3 +1394,92 @@ CREATE SEQUENCE db2.seq2 OWNED BY db1.t.a
 
 statement error invalid OWNED BY option
 ALTER SEQUENCE db2.seq2 OWNED BY doesntexist
+
+
+# Test that passing a descriptor other than a sequence
+# returns an appropriate error.
+subtest invalid_ids
+
+statement ok
+CREATE TABLE t (i SERIAL PRIMARY KEY)
+
+let $t_id
+SELECT 't'::regclass::int
+
+statement error pgcode 42809 "test.public.t" is not a sequence
+SELECT nextval($t_id::regclass)
+
+statement error pgcode 42809 "test.public.t" is not a sequence
+SELECT setval($t_id::regclass, 30, false)
+
+statement error pgcode 42809 "test.public.t" is not a sequence
+SELECT currval($t_id::regclass)
+
+statement ok
+CREATE VIEW v AS (SELECT i FROM t)
+
+let $v_id
+SELECT 'v'::regclass::int
+
+statement error pgcode 42809 "test.public.v" is not a sequence
+SELECT nextval($v_id::regclass)
+
+statement error pgcode 42809 "test.public.v" is not a sequence
+SELECT setval($v_id::regclass, 30, false)
+
+statement error pgcode 42809 "test.public.v" is not a sequence
+SELECT currval($v_id::regclass)
+
+statement ok
+CREATE SCHEMA sc
+
+let $sc_id
+SELECT id FROM system.namespace WHERE name='sc'
+
+statement error does not exist
+SELECT nextval($sc_id::regclass)
+
+statement error does not exist
+SELECT setval($sc_id::regclass, 30, false)
+
+statement error does not exist
+SELECT currval($sc_id::regclass)
+
+statement ok
+CREATE DATABASE db
+
+let $db_id
+SELECT id FROM system.namespace WHERE name='db'
+
+statement error does not exist
+SELECT nextval($db_id::regclass)
+
+statement error does not exist
+SELECT setval($db_id::regclass, 30, false)
+
+statement error does not exist
+SELECT currval($db_id::regclass)
+
+statement ok
+CREATE TYPE e AS ENUM ('foo', 'bar')
+
+let $e_id
+SELECT id FROM system.namespace WHERE name='e'
+
+statement error does not exist
+SELECT nextval($e_id::regclass)
+
+statement error does not exist
+SELECT setval($e_id::regclass, 30, false)
+
+statement error does not exist
+SELECT currval($e_id::regclass)
+
+statement error does not exist
+SELECT nextval(12345::regclass) # Bogus ID
+
+statement error does not exist
+SELECT setval(12345::regclass, 30, false) # Bogus ID
+
+statement error does not exist
+SELECT currval(12345::regclass) # Bogus ID

--- a/pkg/sql/sequence.go
+++ b/pkg/sql/sequence.go
@@ -29,6 +29,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/sql/privilege"
 	"github.com/cockroachdb/cockroach/pkg/sql/sem/builtins"
 	"github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
+	"github.com/cockroachdb/cockroach/pkg/sql/sqlerrors"
 	"github.com/cockroachdb/cockroach/pkg/util/errorutil/unimplemented"
 	"github.com/cockroachdb/cockroach/pkg/util/log"
 	"github.com/cockroachdb/cockroach/pkg/util/sequence"
@@ -94,6 +95,13 @@ func (p *planner) IncrementSequenceByID(ctx context.Context, seqID int64) (int64
 	descriptor, err := p.Descriptors().GetImmutableTableByID(ctx, p.txn, descpb.ID(seqID), flags)
 	if err != nil {
 		return 0, err
+	}
+	if !descriptor.IsSequence() {
+		seqName, err := p.getQualifiedTableName(ctx, descriptor)
+		if err != nil {
+			return 0, err
+		}
+		return 0, sqlerrors.NewWrongObjectTypeError(seqName, "sequence")
 	}
 	return incrementSequenceHelper(ctx, p, descriptor)
 }
@@ -178,6 +186,9 @@ func (p *planner) GetLatestValueInSessionForSequenceByID(
 	if err != nil {
 		return 0, err
 	}
+	if !descriptor.IsSequence() {
+		return 0, sqlerrors.NewWrongObjectTypeError(seqName, "sequence")
+	}
 	return getLatestValueInSessionForSequenceHelper(p, descriptor, seqName)
 }
 
@@ -229,6 +240,9 @@ func (p *planner) SetSequenceValueByID(
 	seqName, err := p.getQualifiedTableName(ctx, descriptor)
 	if err != nil {
 		return err
+	}
+	if !descriptor.IsSequence() {
+		return sqlerrors.NewWrongObjectTypeError(seqName, "sequence")
 	}
 	return setSequenceValueHelper(ctx, p, descriptor, newVal, isCalled, seqName)
 }

